cpp/dbscan.cpp

```diff
@@ -1,183 +1,138 @@
-        }
+Dbscan::Dbscan(float const eps, std::uint32_t const min_samples, std::size_t const num_points_hint)
+    : eps_{eps * eps}
+    , min_samples_{min_samples}
+{
+    if (num_points_hint > 0) {
+        labels_.reserve(num_points_hint);
+        neighbors_.reserve(num_points_hint);
+        visited_.reserve(num_points_hint);
+        to_visit_.reserve(num_points_hint);
+    }
+}
 
-    private:
-        py::detail::unchecked_reference<T, num_dims> points_;
-    };
+auto Dbscan::fit_predict(std::vector<Dbscan::Point> const& points) -> std::vector<Dbscan::Label>
+{
+    PointsVectorAdaptor adapter{points};
 
-public:
-    static constexpr std::int32_t undefined{-2};
-    static constexpr std::int32_t noise{-1};
+    constexpr auto num_dims{2};
+    constexpr auto leaf_size{32};
+    nanoflann::
+        KDTreeSingleIndexAdaptor<nanoflann::L2_Adaptor<float, PointsVectorAdaptor>, PointsVectorAdaptor, num_dims>
+            points_kd_tree{num_dims, adapter, nanoflann::KDTreeSingleIndexAdaptorParams{leaf_size}};
+    points_kd_tree.buildIndex();
 
-    static constexpr auto num_dims{2};
+    nanoflann::SearchParams params{};
+    params.sorted = false;
 
-    DbScan(float eps, std::uint32_t min_samples)
-        : eps_{eps * eps}
-        , min_samples_{min_samples}
-    {
-    }
+    labels_.clear();
+    labels_.resize(std::size(points), undefined);
 
-    auto fit_predict(py::array_t<float> const& points_in)
-    {
-        ScopedTimer timer{};
-        using Adapter = ArrayVectorAdaptor<float, num_dims>;
-        Adapter adapter{points_in};
+    to_visit_.clear();
+    visited_.clear();
 
-        auto const points{points_in.unchecked<num_dims>()};
+    std::int32_t cluster_count{0};
+    std::vector<std::int32_t> clusters{};
 
-        nanoflann::KDTreeSingleIndexAdaptor<nanoflann::L2_Adaptor<float, Adapter>, Adapter, num_dims> points_kd_tree{
-            num_dims, adapter, nanoflann::KDTreeSingleIndexAdaptorParams{32}};
-        points_kd_tree.buildIndex();
+    for (auto i{0UL}; i < std::size(points); ++i) {
+        // skip point if it has already been processed
+        if (labels_[i] != undefined) {
+            continue;
+        }
 
-        std::cout << "building index: " << (static_cast<double>(timer.elapsed_ns()) / 1000.0) << " µs" << std::endl;
+        // find number of neighbors of current point
+        if (points_kd_tree.radiusSearch(points[i].data(), eps_, neighbors_, params) < min_samples_) {
+            labels_[i] = noise;
+            continue;
+        }
 
-        nanoflann::SearchParams params{};
-        params.sorted = false;
+        // This point has at least min_samples_ in its eps neighborhood, so it's considered a core point. Time to
+        // start a new cluster.
 
-        labels_.clear();
-        labels_.resize(points.shape(0), undefined);
+        auto const current_cluster_id{cluster_count++};
+        labels_[i] = current_cluster_id;
 
         to_visit_.clear();
         visited_.clear();
```

Comment:
The size of `points `does not change during this loop if I'm not mistaken. Maybe exchange the `clear() `and `resize()` calls with an `assign()`? Would improve readability in my opinion.

Reply:
`assign` is better here. I changed it.